lib/utils.h

```diff
@@ -331,4 +331,18 @@ static inline const char* bool_to_string(bool b)
     return b ? "true" : "false";
 }
 
+namespace cppcheck
+{
+    NORETURN inline void unreachable()
+    {
+#if defined(__GNUC__)
+        __builtin_unreachable();
```

Comment:
what does these __builtin_unreachable and __assume do? I fear that for instance the GUI will just quit suddenly not telling the user anything about what the problem is. What would the drawback be to throw an internal error?

Reply:
This is code which will never be executed. It is just to appease compilers which do not realize this (GCC and MSVC warn but Clang doesn't).

Reply:
It is essentially the same as `[[noreturn]]`. It is a hint for the compiler that the code can never be reached.

Reply:
But if it does get called. What happens? Is that undefined behavior then?

What if Cppcheck has a bug that means a enum will not have one of the possible values..

A throw and internal error would be nicer than dragons flying out of the nose.. :-)

Reply:
hmm on one hand I fully agree that the code can't be executed.

on the other hand.. if there would be a bug so a enum would have an impossible value then it would be great with an internal error. I fear that these functions might just terminate with no indication what happened.

Reply:
You could argue this for any part of code as this triggers UB. Also this is a known pattern in lots of project and a way to handle these "false positive" compiler warnings (I got this approach from the LLVM code).

Having an impossible value is either impossible, requires malicious intend or known bad (i.e. fuzzing) input and as we have argued several times in the past the code is not intended to be treated this gracefully in 100% of cases as we would be only fixing such issues and not improve anything.

Reply:
alright